include/dgl/heterograph_interface.h

```diff
@@ -0,0 +1,354 @@
+
+  /*!
+   * \brief Get the out degree of the given vertex.
+   * \note The given vertex should belong to the source vertex type
+   *       of the given edge type.
+   * \param etype The edge type
+   * \param vid The vertex id.
+   * \return the out degree
+   */
+  virtual uint64_t OutDegree(dgl_type_t etype, dgl_id_t vid) const = 0;
+
+  /*!
+   * \brief Get the out degrees of the given vertices.
+   * \note The given vertex should belong to the source vertex type
+   *       of the given edge type.
+   * \param etype The edge type
+   * \param vid The vertex id array.
+   * \return the out degree array
+   */
+  virtual DegreeArray OutDegrees(dgl_type_t etype, IdArray vids) const = 0;
+
+  /*!
+   * \brief Return the successor vector
+   * \note The given vertex should belong to the source vertex type
+   *       of the given edge type.
+   * \param vid The vertex id.
+   * \return the successor vector iterator pair.
+   */
+  virtual DGLIdIters SuccVec(dgl_type_t etype, dgl_id_t vid) const = 0;
+
+  /*!
+   * \brief Return the out edge id vector
+   * \note The given vertex should belong to the source vertex type
+   *       of the given edge type.
+   * \param vid The vertex id.
+   * \return the out edge id vector iterator pair.
+   */
+  virtual DGLIdIters OutEdgeVec(dgl_type_t etype, dgl_id_t vid) const = 0;
+
+  /*!
+   * \brief Return the predecessor vector
+   * \note The given vertex should belong to the dest vertex type
+   *       of the given edge type.
+   * \param vid The vertex id.
+   * \return the predecessor vector iterator pair.
+   */
+  virtual DGLIdIters PredVec(dgl_type_t etype, dgl_id_t vid) const = 0;
+
+  /*!
+   * \brief Return the in edge id vector
+   * \note The given vertex should belong to the dest vertex type
+   *       of the given edge type.
+   * \param vid The vertex id.
+   * \return the in edge id vector iterator pair.
+   */
+  virtual DGLIdIters InEdgeVec(dgl_type_t etype, dgl_id_t vid) const = 0;
+
+  /*!
+   * \brief Reset the data in the graph and move its data to the returned graph object.
+   * \return a raw pointer to the graph object.
+   */
+  virtual HeteroGraphInterface *Reset() = 0;
+
+  /*!
+   * \brief Get the adjacency matrix of the graph.
+   *
+   * By default, a row of returned adjacency matrix represents the destination
+   * of an edge and the column represents the source.
+   *
+   * If the fmt is 'csr', the function should return three arrays, representing
+   *  indptr, indices and edge ids
+   *
+   * If the fmt is 'coo', the function should return one array of shape (2, nnz),
+   * representing a horitonzal stack of row and col indices.
+   *
+   * \param transpose A flag to transpose the returned adjacency matrix.
+   * \param fmt the format of the returned adjacency matrix.
+   * \return a vector of IdArrays.
+   */
+  virtual std::vector<IdArray> GetAdj(
+      dgl_id_t etype, bool transpose, const std::string &fmt) const = 0;
```

Comment:
should we add a slicing operation? it slices a vew of a certain edge type.

Reply:
Done.